docs/Configuration.md
## Configuration

Skuber supports both out-of-cluster and in-cluster configurations.
Сonfiguration algorithm can be described as follows:

Initially Skuber tries out-of-cluster methods in sequence (stops on first successful):
- Read `SKUBER_URL` environment variable and use it as kubectl proxy url. If not set then:
- Read `SKUBER_CONFIG` environment variable and if is equal to:
    * `file`  - Skuber will read `~/.kube/config` and use it as configuration source
    * `proxy` - Skuber will assume that kubectl proxy running on `localhost:8080` and will use it as endpoint to cluster API 
    * Otherwise treats contents of `SKUBER_CONFIG` as a file path to kubeconfig file (use it if your kube config file is in custom location). If not present then:
- Read `KUBECONFIG` environment variable and use its contents as path to kubconfig file (similar to `SKUBER_CONFIG`)

If none of the above configuration items are set, then Skuber assumes it is running inside a pod in the cluster, in which case it can always load the required configuration and credentials from the same location in the local file system (in-cluster configuration). If attempting to run your client inside a pod, this will be the preferred configuration.

### Cluster URL

If proxying via a [kubectl proxy](https://kubernetes.io/docs/tasks/extend-kubernetes/http-proxy-access-api/) then you can configure Skuber to connect through that proxy by setting the SKUBER_URL environment variable to point at it e.g.

    export SKUBER_URL=http://localhost:8001

If the cluster URL is set this way, then the SKUBER_CONFIG and KUBECONFIG environment variables below are ignored by Skuber.

Alternatively, you can programmatically configure the proxy details by using one of the [Configuration helper methods](https://github.com/doriordan/skuber/blob/master/client/src/main/scala/skuber/api/Configuration.scala#L32) to construct an appropriate Configuration object to be passed to the `k8sInit` call.

### Kubeconfig file

If not using a `kubectl proxy` then most clients will be configured using a [kubeconfig file](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/). This is the standard configuration file format used by other Kubernetes clients such as `kubectl`. These files allow the following to be configured:

- Authentication credentials - [see below](#security).
- [Namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/) - this allows the client to read/write Kubernetes resources in different cluster namespaces just by changing runtime configuration, which supports partitioning by team / organization.
- Cluster address - i.e the URL to which the client connects to communicate with the Kubernetes API server. This can be either a non-TLS (http) or TLS (https) URL.

Skuber generally follows the same conventions for configuring the client as other Kubernetes client tools such as `kubectl`- to use a specific kubeconfig file, set the `KUBECONFIG` environment variable to the location of the config file e.g.

    export KUBECONFIG=~/.kube_config
 
The use of the kubeconfig format means that `kubectl` can be used to modify configuration settings for Skuber clients, without requiring direct editing of the configuration file - and this is the recommended approach.

Kubeconfig files can contain multiple contexts, each encapsulating the full details required to configure a client - Skuber always configures itself (when initialised by a `k8sInit` call) from the [current context](https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/#define-clusters-users-and-contexts) as set by the `kubectl config use-context`command

Because all of the above configuration items in the configuration file are the same as used by other Kubernetes clients such as kubectl, you (or rather the organization deploying the Skuber application) can share configuration with such other clients or set up separate configuration files for applications depending on organizational security policies, deployment processes and other requirements. 

Note that - unlike the Go language client - Skuber does not attempt to merge different sources of configuration.

*(Configuration can alternatively be passed programmatically to the `k8sInit` call, see the programming guide for details.)*

### Security

When using kubeconfig files, Skuber supports standard security configuration as described below.

If the current context specifies a **TLS** connection (i.e. a `https://` URL) to the cluster server, Skuber will utilise the configured **certificate authority** to verify the server (unless the `insecure-skip-tls-verify` flag is set to true, in which case Skuber will trust the server without verification). 

For client authentication **client certificates** (cert and private key pairs) can be specified for the case where TLS is in use. In addition to client certificates Skuber will use any **bearer token** or **basic authentication** credentials specified. Token or basic auth can be configured as an alternative to or in conjunction with client certificates. 

*(Skuber loads configured server and client certificates / keys directly from the kubeconfig file (or from another location in the file system in the case where the configuration specifies a path rather than embedded data). This means there is no need to store them in the Java trust or key stores.)*


client/src/main/scala/skuber/api/Configuration.scala
package skuber.api

import com.typesafe.config.{ Config, ConfigFactory }

import java.net.URL
import java.time.Instant
import java.time.format.DateTimeFormatter
import scala.collection.JavaConverters._
import scala.util.Try
import scala.util.Failure
import java.util.{ Base64, Date }
import org.yaml.snakeyaml.Yaml
import skuber.{ Namespace, api }
import skuber.api.client._

import scala.concurrent.Future
import scala.concurrent.duration.FiniteDuration
import scala.io.Source
import scala.util.control.NonFatal

/**
 * @author David O'Riordan
 */
case class Configuration(
      clusters: Map[String, Cluster] = Map(),
      contexts: Map[String, Context] = Map(),
      currentContext: Context = Context(),
      users: Map[String, AuthInfo] = Map()) {

  def withCluster(name: String, cluster: Cluster): Configuration = this.copy(clusters = this.clusters + (name -> cluster))

  def withContext(name: String, context: Context): Configuration = this.copy(contexts = this.contexts + (name -> context))

  def useContext(context: Context): Configuration = this.copy(currentContext = context)

  def setCurrentNamespace(namespaceName: String): Configuration =
    this.copy(currentContext = this.currentContext.copy(namespace = Namespace.forName(namespaceName)))

}

object Configuration {

  // local proxy default config is suitable for use with kubectl proxy running on localhost:8080
  lazy val useLocalProxyDefault: Configuration = {
    val defaultCluster=Cluster()
    val defaultContext=Context(cluster=defaultCluster)
    Configuration(
      clusters = Map("default" -> defaultCluster),
      contexts= Map("default" -> defaultContext),
      currentContext = defaultContext)
  }

  // This covers most of RFC3339
  private val DateFormatters = List(DateTimeFormatter.ISO_INSTANT, DateTimeFormatter.ISO_OFFSET_DATE_TIME)
  private def parseInstant(s: String): Instant = {
    DateFormatters.collectFirst(Function.unlift(format => Try(Instant.from(format.parse(s))).toOption))
      .getOrElse(sys.error(s"'$s' could not be parsed as a date time string"))
  }

  // config to use a local proxy running on a specified port
  def useLocalProxyOnPort(port: Int): Configuration = {
    val clusterAddress=s"http://localhost:${port.toString}"
    val defaultCluster=Cluster(server = clusterAddress)
    val defaultContext=Context(cluster=defaultCluster)
    Configuration(
      clusters = Map("default" -> defaultCluster),
      contexts= Map("default" -> defaultContext),
      currentContext = defaultContext)
  }

  // config to use a proxy at specified address
  def useProxyAt(proxyAddress: String): Configuration = {
    val clusterAddress=proxyAddress
    val defaultCluster=Cluster(server = clusterAddress)
    val defaultContext=Context(cluster=defaultCluster)
    Configuration(
      clusters = Map("default" -> defaultCluster),
      contexts= Map("default" -> defaultContext),
      currentContext = defaultContext)
  }

  /**
     * Parse a kubeconfig file to get a K8S Configuration object for the API.
     *
     * See https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/kubeconfig-file.md
     * for format of the kubeconfig file.
     * Enables sharing of config with kubectl when skuber client is not simply directed via a kubectl proxy
     * However note that the merging functionality described at the link above is not implemented
     * in the Skuber library.
    **/
    import java.nio.file.{Files, Path, Paths}
    def parseKubeconfigFile(path: Path = Paths.get(System.getProperty("user.home"),".kube", "config")) : Try[Configuration] = {
      Try {
        val kubeconfigDir: Path = path.getParent
        val is = Files.newInputStream(path)
        (is,kubeconfigDir)
      } flatMap { case (is, kubeconfigDir) =>
        parseKubeconfigStream(is, Some(kubeconfigDir))
      }
    }

    def parseKubeconfigStream(is: java.io.InputStream, kubeconfigDir: Option[Path] = None) : Try[Configuration]= {

      type YamlMap = java.util.Map[String, Object]
      type TopLevelYamlList = java.util.List[YamlMap]

      Try {
        val yaml = new Yaml()
        val mainConfig = yaml.load(is).asInstanceOf[YamlMap]

        def name(parent: YamlMap) =
          parent.get("name").asInstanceOf[String]

        def child(parent: YamlMap, key: String) =
          parent.get(key).asInstanceOf[YamlMap]

        def topLevelList(key: String) =
          mainConfig.get(key).asInstanceOf[TopLevelYamlList]

        def valueAt[T](parent: YamlMap, key: String, fallback: Option[T] = None) : T =
          parent.asScala.get(key).orElse(fallback).get.asInstanceOf[T]

        def optionalInstantValueAt[T](parent: YamlMap, key: String) : Option[Instant] =
          parent.asScala.get(key).flatMap {
            case d: Date => Some(d.toInstant)
            case s: String => Try(parseInstant(s)).toOption
            case _ => None
          }

        def optionalValueAt[T](parent: YamlMap, key: String) : Option[T] =
          parent.asScala.get(key).map(_.asInstanceOf[T])

        def pathOrDataValueAt[T](parent: YamlMap, pathKey: String, dataKey: String) : Option[PathOrData] = {
          val path = optionalValueAt[String](parent, pathKey)
          val data = optionalValueAt[String](parent, dataKey)

          // Return some Right if data value is set, otherwise some Left if path value is set
          // if neither is set return None
          // Note - implication is that a data setting overrides a path setting
          (path, data) match {
            case (_, Some(b64EncodedData)) => Some(Right(Base64.getDecoder.decode(b64EncodedData)))
            case (Some(p), _) =>
              // path specified
              // if it is a relative path and a directory was specified then construct full path from those components,
              // otherwise just return path as given
              val expandedPath = (Paths.get(p), kubeconfigDir) match {
                case (basePath, Some(dir)) if !basePath.isAbsolute =>
                  Paths.get(dir.normalize.toString, basePath.normalize.toString).normalize.toString
                case _  => p
              }
              Some(Left(expandedPath))
            case (None, None) => None
          }
        }

        def topLevelYamlToK8SConfigMap[K8SConfigKind](kind: String, toK8SConfig: YamlMap=> K8SConfigKind) =
          topLevelList(kind + "s").asScala.map(item => name(item) -> toK8SConfig(child(item, kind))).toMap


        def toK8SCluster(clusterConfig: YamlMap) =
          Cluster(
            apiVersion=valueAt(clusterConfig, "api-version", Some("v1")),
            server=valueAt(clusterConfig,"server",Some("http://localhost:8001")),
            insecureSkipTLSVerify=valueAt(clusterConfig,"insecure-skip-tls-verify",Some(false)),
            certificateAuthority=pathOrDataValueAt(clusterConfig, "certificate-authority","certificate-authority-data")
          )


        val k8sClusterMap = topLevelYamlToK8SConfigMap("cluster", toK8SCluster)

        def toK8SAuthInfo(userConfig:YamlMap): AuthInfo = {

          def authProviderRead(authProvider: YamlMap): Option[AuthProviderAuth] = {
            val config = child(authProvider, "config")
            name(authProvider).toLowerCase match {
              case "oidc" =>
                Some(OidcAuth(idToken = valueAt(config, "id-token")))
              case "gcp" =>
                Some(
                  GcpAuth(
                    accessToken = optionalValueAt(config, "access-token"),
                    expiry = optionalInstantValueAt(config, "expiry"),
                    cmdPath = valueAt(config, "cmd-path"),
                    cmdArgs = valueAt(config, "cmd-args")
                  )
                )
              case _ => None
            }
          }

          val maybeAuth = optionalValueAt[YamlMap](userConfig, "auth-provider") match {
            case Some(authProvider) => authProviderRead(authProvider)
            case None =>
              val clientCertificate = pathOrDataValueAt(userConfig, "client-certificate", "client-certificate-data")
              val clientKey = pathOrDataValueAt(userConfig, "client-key", "client-key-data")

              val token = optionalValueAt[String](userConfig, "token")

              val userName = optionalValueAt[String](userConfig, "username")
              val password = optionalValueAt[String](userConfig, "password")

              (userName, password, token, clientCertificate, clientKey) match {
                case (Some(u), Some(p), _, _, _) => Some(BasicAuth(u, p))
                case (_, _, Some(t), _, _) => Some(TokenAuth(t))
                case (u, _, _, Some(cert), Some(key)) => Some(CertAuth(cert, key, u))
                case _ => None
              }
          }

          maybeAuth.getOrElse(NoAuth)
        }
        val k8sAuthInfoMap = topLevelYamlToK8SConfigMap("user", toK8SAuthInfo)

        def toK8SContext(contextConfig: YamlMap) = {
          val cluster=contextConfig.asScala.get("cluster").filterNot(_.asInstanceOf[String] == "").flatMap { clusterName =>
            k8sClusterMap.get(clusterName.asInstanceOf[String])
          }.getOrElse(Cluster())
          val authInfo =contextConfig.asScala.get("user").filterNot(_.asInstanceOf[String] == "").flatMap { userKey =>
            k8sAuthInfoMap.get(userKey.asInstanceOf[String])
          }.getOrElse(NoAuth)
          val namespace=contextConfig.asScala.get("namespace").fold(Namespace.default) { name=>Namespace.forName(name.asInstanceOf[String]) }
          Context(cluster,authInfo,namespace)
        }

        val k8sContextMap = topLevelYamlToK8SConfigMap("context", toK8SContext _)

        val currentContextStr: Option[String] = optionalValueAt(mainConfig, "current-context")
        val currentContext = currentContextStr.flatMap(k8sContextMap.get(_)).getOrElse(Context())

        Configuration(k8sClusterMap, k8sContextMap, currentContext, k8sAuthInfoMap)
      }
    }

  private lazy val inClusterConfigReloadInterval: Option[FiniteDuration] = {
    import scala.concurrent.duration._
    // We default to 10 minutes because modern Kubernetes distributions issue service account tokens that are valid for
    // one hour, rotating them when they reach 80% of their lifespan. This means, any given arbitrary read of the token
    // from the filesystem can only be sure that that token will be valid for a maximum of 12 minutes (20% of 1 hour).
    // This is configurable, hence we allow it to be overridden by an environment variable.
    // Note that the equivalent Go code for this reloads every minute, erroneously stating in a comment that tokens are
    // rotated every 10 minutes. See https://github.com/kubernetes/client-go/blob/4f9edc15a7e71c3f9c7874a872a2545c8737726c/transport/token_source.go#L75-L79
    // If the value is less or equal to zero, then we don't reload.
    sys.env.get("SKUBER_TOKEN_RELOAD_INTERVAL_SECONDS")
      .map(_.toInt)
      .orElse(Some(600))
      .filter(_ > 0)
      .map(_.seconds)
  }

  /**
    * Tries to create in-cluster configuration using credentials mounted inside a running pod
    * <p>Follows official golang client logic
    *
    * @return Try[Configuration]
    *
    * @see https://kubernetes.io/docs/tasks/access-application-cluster/access-cluster/#accessing-the-api-from-a-pod
    *      https://github.com/kubernetes/client-go/blob/master/rest/config.go#L313
    *      https://github.com/kubernetes-client/java/blob/master/util/src/main/java/io/kubernetes/client/util/ClientBuilder.java#L134
    */
  lazy val inClusterConfig: Try[Configuration] = {

    val rootK8sFolder = "/var/run/secrets/kubernetes.io/serviceaccount"
    val tokenPath     = s"$rootK8sFolder/token"
    val namespacePath = s"$rootK8sFolder/namespace"
    val caPath        = s"$rootK8sFolder/ca.crt"

    lazy val maybeHost: Try[String] = Try(sys.env("KUBERNETES_SERVICE_HOST"))
      .recoverWith { case e: NoSuchElementException =>
        Failure(new Exception("environment variable KUBERNETES_SERVICE_HOST must be defined", e))}

    lazy val maybePort: Try[String] = Try(sys.env("KUBERNETES_SERVICE_PORT"))
      .recoverWith { case e: NoSuchElementException =>
        Failure(new Exception("environment variable KUBERNETES_SERVICE_PORT must be defined", e))}

    def tryLoadToken() = tryReadPath(tokenPath)

    lazy val maybeNamespace = tryReadPath(namespacePath)

    // is not strictly required
    // but client-go tries to read ca.file and logs the following error if unable to and continues
    //"Expected to load root CA config from %s, but got err: %v", rootCAFile, err)
    lazy val ca: Option[PathOrData] = if (Files.exists(Paths.get(caPath))) Some(Left(caPath)) else None

    for {
      host      <- maybeHost
      port      <- maybePort
      token     <- tryLoadToken()
      namespace <- maybeNamespace
    } yield {
      val hostPort = s"https://$host${if (port.nonEmpty) ":" + port else ""}"
      val cluster = Cluster(server = hostPort, certificateAuthority = ca)
      val auth = inClusterConfigReloadInterval match {
        case Some(reloadInterval) =>
          reloadableAccessTokenAuth { () =>
            Future.fromTry(tryLoadToken().map(t => (reloadInterval, t)))
          }
        case None =>
          TokenAuth(token)
      }
      val ctx = Context(cluster, auth, Namespace.forName(namespace))

      Configuration(
        clusters = Map("default" -> cluster),
        contexts = Map("default" -> ctx),
        currentContext = ctx
      )
    }
  }

  private def tryReadPath(path: String): Try[String] = Try {
    val source = Source.fromFile(path, "utf-8")
    try {
      source.getLines().mkString("\n")
    } finally {
      try {
        source.close()
      } catch {
        case NonFatal(_) =>
          // Ignore
      }
    }
  }

  /*
   * Get the current default configuration
   */
  def defaultK8sConfig: Configuration = {
    import java.nio.file.Paths

    val skuberUrlOverride = sys.env.get("SKUBER_URL")
    skuberUrlOverride match {
      case Some(url) =>
        Configuration.useProxyAt(url)
      case None =>
        val skuberConfigEnv = sys.env.get("SKUBER_CONFIG")
        skuberConfigEnv match {
          case Some(conf) if conf == "file" =>
            Configuration.parseKubeconfigFile().get // default kubeconfig location
          case Some(conf) if conf == "proxy" =>
            Configuration.useLocalProxyDefault
          case Some(fileUrl) =>
            val path = Paths.get(new URL(fileUrl).toURI)
            Configuration.parseKubeconfigFile(path).get
          case None =>
            // try KUBECONFIG
            val kubeConfigEnv = sys.env.get("KUBECONFIG")
            kubeConfigEnv.map { kc =>
              Configuration.parseKubeconfigFile(Paths.get(kc))
            }.getOrElse {
              // Try to get config from a running pod
              // if that is not set then use default kubeconfig location
              Configuration.inClusterConfig.orElse(
                Configuration.parseKubeconfigFile()
              )
            }.get
        }
    }
  }

}
